lib/Sema/CSDiag.cpp

```diff
@@ -2883,7 +2887,8 @@ static bool addTypeCoerceFixit(InFlightDiagnostic &diag, ConstraintSystem &CS,
     diag.fixItInsert(
         Lexer::getLocForEndOfToken(CS.DC->getASTContext().SourceMgr,
                                    expr->getEndLoc()),
-        (llvm::Twine(canUseAs ? " as " : " as! ") + OS.str()).str());
+        (llvm::Twine(canUseAs ? " as " : " as! ") + OS.str() +
+         llvm::Twine(canUseAs && bothOptional ? "?" : "")).str());
```

Comment:
This is a little iffy; not all types can just have `?` stuck on the end. (Consider `() -> Void`.) It'd be better to add one layer back in the both-optional case.

Reply:
Makes sense..

Reply:
Wait but both types already have to be optional to produce this fix, right?

Reply:
Sorry what I mean is - if both types were optional before they could be again right?

Reply:
Doesn't `() -> Void` just need parens around it to be optional e.g. `(() -> Void)?`

Reply:
Right. Before Greg was adding the `?` manually, but I pointed out that sometimes that's not good enough. So now he's rebuilding the optional type and having it print itself.

Reply:
Ah sorry for the noise, I totally misunderstand...